app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,19 @@
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  def facebook
+    # You need to implement the method below in your model (e.g. app/models/user.rb)
+    @user = User.from_omniauth(request.env["omniauth.auth"])
+
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
+      set_flash_message(:notice, :success, :kind => "Facebook") if is_navigational_format?
+      self.current_user = @user
+    else
+      session["devise.facebook_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_url
```

Comment:
Can we change the logic here to not redirect to the registration form? I think the best logic will be in the `User.from_omniauth`:

1. Search for the `provider` and `uid` in the user table
2. If user can't be found, search for a user with empty `provider` and `uid` and with the same `email`. If found update the `provider` and `uid` and login with that user
3. If can't find user with that email, create a new one